File: DSA/SuryanshMishra_590012069/Day93/day93question3.cpp
#include <vector>
#include <algorithm>
#include <cstddef>

using namespace std;

// DP: minimum coins to make amount
int coinChange(const vector<int>& coins, int amount) {
    const int INF = amount + 1;
    vector<int> dp(amount+1, INF);
    dp[0] = 0;
    for (int a = 1; a <= amount; ++a) {
        for (int c : coins) {
            if (a - c >= 0) {
                dp[a] = min(dp[a], dp[a-c] + 1);
            }
        }
    }
    return dp[amount] == INF ? -1 : dp[amount];
}

File: DSA/VaniAgrawal_590015846/Day_93_Question_3.py
# Coin Change (Min Coins)
def coin_change(coins, amount):
    dp = [float('inf')] * (amount+1)
    dp[0] = 0
    for coin in coins:
        for x in range(coin, amount+1):
            dp[x] = min(dp[x], dp[x-coin] + 1)
    return -1 if dp[amount] == float('inf') else dp[amount]

coins = list(map(int, input().split()))
amount = int(input())
print(coin_change(coins, amount))
